File: server/app/schemas/sessions.py
"""세션(운동 로그) Pydantic 스키마."""

from datetime import datetime

from pydantic import BaseModel, Field


# ── 세션 생성 ─────────────────────────────────────────────────────────────────
class StartSessionRequest(BaseModel):
    routine_id: str | None = None
    routine_day_id: str | None = None
    gym_id: str | None = None


class SessionStartData(BaseModel):
    session_id: str
    routine_id: str | None = None
    routine_name: str | None = None
    gym_id: str | None = None
    started_at: datetime
    message: str = "운동을 시작합니다!"


class SessionData(BaseModel):
    session_id: str
    routine_day_id: str | None = None
    gym_id: str | None = None
    started_at: datetime
    finished_at: datetime | None = None
    status: str
    routine_name: str | None = None
    duration_minutes: int | None = None
    total_sets: int | None = None
    completed_exercises: int | None = None
    total_calories: int | None = None


# ── 세트 기록 ─────────────────────────────────────────────────────────────────
class LogSetRequest(BaseModel):
    exercise_id: str
    routine_exercise_id: str | None = None
    set_number: int = Field(ge=1)
    weight_kg: float | None = None
    reps: int = Field(ge=0)
    rpe: float | None = Field(default=None, ge=1.0, le=10.0)
    is_completed: bool = True


class WorkoutSetItem(BaseModel):
    set_id: str
    exercise_id: str
    exercise_name: str | None = None
    set_number: int
    weight_kg: float | None = None
    reps: int
    rpe: float | None = None
    is_completed: bool
    performed_at: datetime


# ── 세션 종료 ─────────────────────────────────────────────────────────────────
class FinishSessionRequest(BaseModel):
    finished_at: datetime | None = None


# ── 세션 상세 ─────────────────────────────────────────────────────────────────
class SessionDetail(SessionData):
    sets: list[WorkoutSetItem] = Field(default_factory=list)
    total_volume_kg: float = 0.0


class SessionCalendarItem(BaseModel):
    date: str  # YYYY-MM-DD
    session_id: str
    routine_name: str | None = None
    duration_minutes: int | None = None


class SessionCalendarData(BaseModel):
    year: int
    month: int
    records: list[SessionCalendarItem]
    total_session_count: int


class SessionListData(BaseModel):
    items: list[SessionData]


# ── 통계 ──────────────────────────────────────────────────────────────────────
class RecentSessionItem(BaseModel):
    session_id: str
    routine_name: str | None = None
    date: str  # YYYY-MM-DD


class GymStatItem(BaseModel):
    gym_id: str
    gym_name: str
    session_count: int
    total_volume_kg: float


class SessionStatsData(BaseModel):
    total_sessions: int
    total_volume_kg: float
    total_duration_minutes: int
    total_sets: int = 0
    weekly_session_count: int = 0
    streak_days: int
    recent_session: RecentSessionItem | None = None
    by_gym: list[GymStatItem] = Field(default_factory=list)


class VolumeAnalysisItem(BaseModel):
    date: str  # YYYY-MM-DD
    volume_kg: float


class VolumeAnalysisData(BaseModel):
    items: list[VolumeAnalysisItem]


# ── 근육 볼륨 분석 ───────────────────────────────────────────────────────────
class MuscleVolumeItem(BaseModel):
    muscle: str  # MuscleGroup.name_ko
    weekly_volume: float
    optimal_min: float
    optimal_max: float
    status: str  # "OPTIMAL" | "LOW" | "HIGH"


class MuscleVolumeData(BaseModel):
    period: str  # "WEEK" | "MONTH"
    volume_by_muscle: list[MuscleVolumeItem]
    ai_coach_message: str


# ── 휴식 타이머 ──────────────────────────────────────────────────────────────
class RestTimerData(BaseModel):
    rest_seconds: int
    min_rest_seconds: int
    max_rest_seconds: int
    message: str
    based_on: str  # "routine" | "goal_default"

File: app/src/services/sessions.ts
import { apiFetch } from './api';

export interface SessionCalendarItem {
  date: string;
  session_id: string;
  routine_id: string | null;
  routine_name: string | null;
  fitness_goals: string[];
  duration_minutes: number | null;
  gym_name: string | null;
  total_volume_kg: number;
  total_weight_kg: number;
  total_sets: number;
}

export interface SessionCalendarData {
  year: number;
  month: number;
  records: SessionCalendarItem[];
  total_session_count: number;
}

export interface RecentSessionItem {
  session_id: string;
  routine_name: string | null;
  date: string;
}

export interface SessionStatsData {
  total_sessions: number;
  total_volume_kg: number;
  total_weight_kg: number;
  total_duration_minutes: number;
  total_sets: number;
  weekly_session_count: number;
  streak_days: number;
  total_calories_kcal: number;
  recent_session: RecentSessionItem | null;
}

export function getSessions(token: string, year?: number, month?: number): Promise<SessionCalendarData> {
  const params = year != null && month != null ? `?year=${year}&month=${month}` : '';
  return apiFetch<SessionCalendarData>(`/api/v1/sessions${params}`, { token });
}

export function getSessionStats(token: string): Promise<SessionStatsData> {
  return apiFetch<SessionStatsData>('/api/v1/sessions/stats', { token });
}

// ── 세션 시작 / 세트 기록 / 완료 ─────────────────────────────────────────────

export interface StartSessionData {
  session_id: string;
  routine_id: string | null;
  routine_name: string | null;
  gym_id: string | null;
  started_at: string;
}

export interface StartSessionBody {
  routine_id?: string;
  routine_day_id?: string;
  gym_id?: string;
}

export interface LogSetBody {
  exercise_id: string;
  routine_exercise_id?: string;
  set_number: number;
  weight_kg?: number | null;
  reps: number;
  is_completed: boolean;
}

export function startSession(token: string, body: StartSessionBody): Promise<StartSessionData> {
  return apiFetch<StartSessionData>('/api/v1/sessions', {
    token,
    method: 'POST',
    body: JSON.stringify(body),
  });
}

export function logSet(token: string, session_id: string, body: LogSetBody): Promise<unknown> {
  return apiFetch<unknown>(`/api/v1/sessions/${session_id}/sets`, {
    token,
    method: 'POST',
    body: JSON.stringify(body),
  });
}

export function finishSession(token: string, session_id: string): Promise<unknown> {
  return apiFetch<unknown>(`/api/v1/sessions/${session_id}/finish`, {
    token,
    method: 'PATCH',
    body: JSON.stringify({}),
  });
}

// ── 볼륨 분석 / 근육 부위별 분석 ─────────────────────────────────────────────

export interface VolumeAnalysisItem {
  date: string;       // YYYY-MM-DD
  volume_kg: number;
}

export interface VolumeAnalysisData {
  items: VolumeAnalysisItem[];
}

export interface MuscleVolumeItem {
  muscle: string;         // MuscleGroup.name_ko (예: "가슴", "광배근")
  weekly_volume: number;
  optimal_min: number;
  optimal_max: number;
  status: 'OPTIMAL' | 'LOW' | 'HIGH';
}

export interface MuscleVolumeData {
  period: string;
  volume_by_muscle: MuscleVolumeItem[];
  ai_coach_message: string;
}

export function getVolumeAnalysis(token: string, days: number): Promise<VolumeAnalysisData> {
  return apiFetch<VolumeAnalysisData>(`/api/v1/sessions/analysis/volume?days=${days}`, { token });
}

export function getMuscleVolumeAnalysis(token: string, period: 'WEEK' | 'MONTH'): Promise<MuscleVolumeData> {
  return apiFetch<MuscleVolumeData>(`/api/v1/sessions/analysis/muscle-volume?period=${period}`, { token });
}
